pkg/client/events/replay_client.go

```diff
@@ -0,0 +1,244 @@
+var _ client.EventsReplayClient[
+	any,
+	observable.ReplayObservable[any],
+] = (*replayClient[any, observable.ReplayObservable[any]])(nil)
+
+// replayClient implements the EventsReplayClient interface for a generic type T,
+// and replay observable for type T.
+type replayClient[T any, U observable.ReplayObservable[T]] struct {
+	// endpointURL is the URL of RPC endpoint which eventsClient subscription
+	// requests will be sent.
+	endpointURL string
+	// queryString is the query string used to subscribe to events of the
+	// desired type.
+	// See: https://docs.cosmos.network/main/learn/advanced/events#subscribing-to-events
+	// and: https://docs.cosmos.network/main/learn/advanced/events#default-events
+	queryString string
+	// eventsClient is the events query client which is used to subscribe to
+	// newly committed block events. It emits an either value which may contain
+	// an error, at most, once and closes immediately after if it does.
+	eventsClient client.EventsQueryClient
+	// eventBytesToTypeDecoder is a function which decodes event subscription
+	// message bytes into the type defined by the EventsReplayClient's generic type
+	// parameter.
+	eventBytesToTypeDecoder func([]byte) (T, error)
+	// latestObsvbls is a replay observable with replay buffer size 1,
+	// which holds the "active latest observable" which is notified when
+	// new events are received by the events query client subscription
+	// created in goPublishEvents. This observable (and the one it emits) closes
+	// when the events bytes observable returns an error and is updated with a
+	// new "active" observable after a new events query subscription is created.
+	latestObsvbls observable.ReplayObservable[U]
+	// latestObsvblsReplayPublishCh is the publish channel for latestObsvbls.
+	// It's used to set latestObsvbls initially and subsequently update it, for
+	// example, when the connection is re-established after erroring.
+	latestObsvblsReplayPublishCh chan<- U
+}
+
+// NewEventsReplayClient creates a new EventsReplayClient from the given
+// dependencies, cometWebsocketURL and subscription query string. It requires a
+// decoder function to be provided which decodes event subscription message
+// bytes into the type defined by the EventsReplayClient's generic type parameter.
+//
+// Required dependencies:
+//   - client.EventsQueryClient
+func NewEventsReplayClient[T any, U observable.ReplayObservable[T]](
+	ctx context.Context,
+	deps depinject.Config,
+	cometWebsocketURL string,
+	queryString string,
+	eventBytesToTypeDecoder func([]byte) (T, error),
+) (client.EventsReplayClient[T, U], error) {
+	// Initialise the mapped client
+	rClient := &replayClient[T, U]{
+		endpointURL:             cometWebsocketURL,
+		queryString:             queryString,
+		eventBytesToTypeDecoder: eventBytesToTypeDecoder,
+	}
+	latestObsvbls,
+		latestObsvblsReplayPublishCh := channel.NewReplayObservable[U](
+		ctx,
+		latestReplayBufferSize,
```

Comment:
I think this should be received as a parameter and should not have a default (i.e. I think we should remove this const).